src/com/company/stack/StackImpl.java

```diff
@@ -1,13 +1,49 @@
 package com.company.stack;
 
-public class StackImpl implements Stack{
+import java.lang.reflect.Array;
+import java.util.Arrays;
+import java.util.EmptyStackException;
+
+public class StackImpl<T> implements Stack{
+    private T[] array;
+
+    public StackImpl() {
+        this.array = (T[]) new Object[0];
+    }
+
     @Override
-    public int pop() {
-        return 0;
+    public T pop() {
+        if (array.length == 0){
+            throw new EmptyStackException();
+        }
+        T x = this.array[this.array.length - 1];
+
+        T[] newArray = (T[]) new Object[this.array.length - 1];
```

Comment:
Why are you creating a new array? It will create a new array every time we perform a pop from the stack and will use a lot of heap memory.
Think about how you can optimize this.